index.rst

```diff
@@ -1,42 +1,198 @@
-.. _explicit-namespaces:
+.. explicit_namespaces documentation master file
 
-Explicit namespaces in import/export
-~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+Explicit namespaces in import/export lists
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 .. extension:: ExplicitNamespaces
-    :shortdesc: Enable using the keyword ``type`` to specify the namespace of
-        entries in imports and exports (:ref:`explicit-namespaces`).
-        Implied by :extension:`TypeOperators` and :extension:`TypeFamilies`.
+    :shortdesc: Enable the use of the keywords ``type`` and ``data`` to specify the precise
+        entities exported from a module and brought into scope by an import declaration 
+        (:ref:`explicit-namespaces`).
 
+    :implied by: :extension:`TypeOperators`, :extension:`TypeFamilies`
     :since: 7.6.1
 
-    Enable use of explicit namespaces in module export lists.
+    Enable use of explicit namespaces in module import and export lists.
 
-In an import or export list, such as ::
+Background
+~~~~~~~~~~
 
-      module M( f, (++) ) where ...
-        import N( f, (++) )
-        ...
+Haskell has two namespaces:
 
-the entities ``f`` and ``(++)`` are *values*. However, with type
-operators it becomes possible to declare
-``(++)`` as a *type constructor*. In that case, how would you export or
-import it?
+* the *type namespace*, including the names of type constructors, type
+  synonyms, type families and type classes;
 
-The :extension:`ExplicitNamespaces` extension allows you to prefix the name of
-a type constructor in an import or export list with "``type``" to
-disambiguate this case, thus: ::
+* the *data namespace*, including the names of term-level values, functions,
+  data constructors and pattern synonyms.
 
-      module M( f, type (++) ) where ...
-        import N( f, type (++) )
-        ...
-      module N( f, type (++) ) where
-        data family a ++ b = L a | R b
+This separation allows us to define data constructors and type
+constructors whose names coincide:
 
-The extension :extension:`ExplicitNamespaces` is implied by
-:extension:`TypeOperators` and (for some reason) by :extension:`TypeFamilies`.
+.. code:: haskell
 
-In addition, with :extension:`PatternSynonyms` you can prefix the name of a
-data constructor in an import or export list with the keyword
-``pattern``, to allow the import or export of a data constructor without
-its parent type constructor.
```

Comment:
We should retain a mention somewhere that `pattern` is still supported with `PatternSynonyms` (e.g. for users who need to support old code or old GHC releases), but that users are encouraged to use `data` instead.